play-json-ops-common/src/main/scala/play/api/libs/json/ops/PlayJsonMacros.scala

```diff
@@ -0,0 +1,343 @@
+package play.api.libs.json.ops
+
+import play.api.libs.json.{Json, Reads}
+
+import scala.language.reflectiveCalls
+import scala.reflect.macros.whitebox
+import language.experimental.macros
+import scala.collection.generic
+
+/**
+ * Provides macros similar to those in [[Json]] but with modified functionality to better handle our common json
+ * parsing use cases, for example missing fields (instead of empty arrays).
+ */
+object PlayJsonMacros extends TolerantContainerFormats {
+
+  /**
+   * Same as [[Json.reads]] but when reading containers such as [[List]], [[Seq]], [[Set]], etc., if the field
+   * is missing will return an empty container instead of failing to validate the model.
```

Comment:
Purely out of curiosity -- we have real-world situations where this is coming up?  I'd like to hope that we're not generating JSON of this form, although I can totally believe that we're receiving some.

Reply:
It's about reading from partner APIs. Some will skip empty fields or sends `null` instead of sending `[]`. This just avoids the need to capture these as `Option[Seq[String]]` and you can instead fold `null` or missing fields into `Seq.empty[String]`.